### PR TITLE
monitoring: Add time-series db plugins + apis to access time-series

### PR DESCRIPTION
**Pull request by anmolbabu**
This patch adds the following:
1. Manager.py :
 a. On load, it pushes the default monitoring configurations to etcd.
 b. It starts the monitoring api manager.
2. APIManager.py:
 a. Starts flask api server as a multiprocessing.process
 b. Exposes 2 apis:
   i. /monitoring/nodes/<node_name>/<resource_name>/stats: to fetch
     stats of <resource_name> of <node_name>.
  ii. /moitoring/nodes/<node_name>/monitored_types: to fetch the list
     of metrics available in the time series db.
3. DefaultsManager.py: Provide monitoring defaults.
4. TimeSeriesDBManager.py:
  a. Provides plugin based framework. This framework allows
     the developers to define as many time-series db driver/plugins as
     he/she wishes but, at a time only 1 is active depending upon config
     file field.
  b. Adds GraphitePlugin to facilitate queries to graphite and also adds
     InfluxDbPlugin which is currently just a place holder to facilitate
     qeries to influxdb in future.
  c. Any time-series db access enaling plugin like GraphitePlugin need to
     extend an abstract class TimeSeriesDBPlugin.
  d. This file also adds a meta class PluginMount which creates anything
     inheriting from TimeSeriesDBPlugin.
5. utils.py:
  a. Singleton Logger which is created only once and can be used for logging
  b. Singleton TendrlConfig which is created only once and can be used for
     utilizing configurations in config file.
  c. Defines a singleton which can be used to decorate a class as singleton
6. EtcdStore.py: Enables access to etcd.

Change-Id: I95174aa8220f6a38419dd3eb9b7c479f26e53e9a
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

**Comment by anmolbabu** [comment]
@shtripat @nthomas-redhat @r0h4n @brainfunked please review this WIP patch.

TODO:
1. Organise TimeSeriesDBManager into multiple different plugin files
2. Add UTs
3. Add tox related stuff
4. Add Installation steps

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    namespace_packages=['tendrl'],
+    url="http://www.redhat.com",
+    author="Anmol Babu",
+    author_email="anbabu@redhat.com",
+    license="LGPL-2.1+",
+    zip_safe=False,
+    install_requires=install_requires,
+    entry_points={
+        'console_scripts': [
+            'tendrl-alerts = tendrl.Manager:main'
```
Comment (shtripat) [inline setup.py]: I feel its better to use all lowercase names for python sources with underscores if required.

File: tendrl/APIManager.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from tendrl.TimeSeriesDBManager import TimeSeriesDBManager
+
+app = Flask(__name__)
+
+@app.route("/monitoring/nodes/<node_name>/<resource_name>/stats")
+def getStats(node_name, resource_name):
+    return TimeSeriesDBManager().get_plugin().get_metric_stats(node_name, resource_name)
+
+
+@app.route("/moitoring/nodes/<node_name>/monitored_types")
```
Comment (shtripat) [inline tendrl/APIManager.py]: spelling of monitoring?

File: tendrl/APIManager.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    def start(self):
+        p = Process(target=self.startServer())
+        p.start()
+        p.join()
+        self.processes.append(p)
+
+
+    def stop(self):
+        try:
+    	    func = request.environ.get('werkzeug.server.shutdown')
```
Comment (shtripat) [inline tendrl/APIManager.py]: What is this werkzeug.server.shutdown ?

File: tendrl/APIManager.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+from flask import Flask, request
```
Comment (shtripat) [inline tendrl/APIManager.py]: I feel all the monitoring related sources could be under tendrl/monitoring package. Looks inline with tendrl/gluster_bridge

File: tendrl/Manager.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import signal
+import tendrl.utils as utils
+import tendrl.APIManager as api_manager
+import tendrl.DefaultsManager as defaults_manager
+
+
+LOG = utils.Logger().getLogger()
```
Comment (shtripat) [inline tendrl/Manager.py]: Make sure logger is inline with PR https://github.com/Tendrl/gluster_bridge/pull/32

File: tendrl/TimeSeriesDBManager.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    def intialize(self):
+        configs = TendrlConfig()
+        self.host = configs.get('general', 'time_series_db_server')
+        self.port = configs.get('general', 'time_series_db_port')
+        self.prefix = 'collectd.'
+
+
+    def get_metric_stats(self, entity_name, metric_name):
+        metric_name = entity_name.replace('.', '_') + "." + metric_name
+        url = 'http://' + self.host + ':' + str(self.port) + '/render?target=' + self.prefix + metric_name + "&format=json"
```
Comment (shtripat) [inline tendrl/TimeSeriesDBManager.py]: Use something like temlapte % (vars) instead of string concatenation

File: tendrl/TimeSeriesDBManager.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        print url
```
Comment (shtripat) [inline tendrl/TimeSeriesDBManager.py]: Remove debug statements

File: tendrl/TimeSeriesDBManager.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        conn = urllib2.urlopen(url)
+        data = conn.read()
+        conn.close()
+        return data
+
+
+    def get_metrics(self, entity_name):
+        url = 'http://' + self.host + ':' + str(self.port) + '/metrics/index.json'
```
Comment (shtripat) [inline tendrl/TimeSeriesDBManager.py]: Same as above

File: tendrl/TimeSeriesDBManager.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        conn = urllib2.urlopen(url)
+        data = conn.read()
+        metrics = ast.literal_eval(data)
+        result = []
+        prefix = 'collectd.' + entity_name.replace('.', '_') + "."
```
Comment (shtripat) [inline tendrl/TimeSeriesDBManager.py]: template % (vars)

File: tendrl/TimeSeriesDBManager.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        conn.close()
+        return str(result)
+
+
+class InfluxDbPlugin(TimeSeriesDBPlugin):
+    def intialize(self):
+        pass
+
+    def get_metric_stats(self, entity_name, metric_name):
+        print "Need to implement metric fetcher for influxdb"
```
Comment (shtripat) [inline tendrl/TimeSeriesDBManager.py]: Better just put a TODO and leave at the moment

File: tendrl/default_values.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import socket
```
Comment (shtripat) [inline tendrl/default_values.py]: This seems to be a configuration. Can we think of making this as yaml/json ?

File: tendrl/utils.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            self.path = DEFAULT_CONFIG_PATH
+        if not os.path.exists(self.path):
+            err = ConfigNotFound("Tendrl Gluster Bridge Configuration not found at %s" % self.path)
+            #LOG.error(err, exc_info=True)
+            raise err
+        self.read(self.path)
+
+
+@singleton
+class Logger():
```
Comment (shtripat) [inline tendrl/utils.py]: Logger to be inline with PR https://github.com/Tendrl/gluster_bridge/pull/32

File: tendrl/APIManager.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+def getStats(node_name, resource_name):
+    return TimeSeriesDBManager().get_plugin().get_metric_stats(node_name, resource_name)
+
+
+@app.route("/moitoring/nodes/<node_name>/monitored_types")
+def getStatTypes(node_name):
+    return TimeSeriesDBManager().get_plugin().get_metrics(node_name)
+
+
+class APIManager():
```
Comment (nthomas-redhat) [inline tendrl/APIManager.py]: As discussed, move this impl to gevent rather than using multiprocessing

File: tendrl/APIManager.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    	    if func is None:
+    		    raise RuntimeError('Not running with the Werkzeug Server')
+            func()
+            for p in self.processes:
+                p.terminate()
+        except Exception as e:
+            LOG.error('Exception %s caught while interrupting api server'.format(str(e)), exc_info=True)
+
+
+def main():
```
Comment (nthomas-redhat) [inline tendrl/APIManager.py]: Remove the main here

File: tendrl/DefaultsManager.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+class DefaultsManager():
+    def pushMonitoringConfigs(self, etcd_dir, configs):
+        EtcdStore().etcd_client.write(etcd_dir, str(configs))
+
+
+    def start(self):
+        self.pushMonitoringConfigs(MONITORING_DEFAULTS_DIRECTORY, default_values.monitoring_config)
+
+
+def main():
```
Comment (nthomas-redhat) [inline tendrl/DefaultsManager.py]: Remove main here

File: tendrl/TimeSeriesDBManager.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        self.host = configs.get('general', 'time_series_db_server')
+        self.port = configs.get('general', 'time_series_db_port')
+        self.prefix = 'collectd.'
+
+
+    def get_metric_stats(self, entity_name, metric_name):
+        metric_name = entity_name.replace('.', '_') + "." + metric_name
+        url = 'http://' + self.host + ':' + str(self.port) + '/render?target=' + self.prefix + metric_name + "&format=json"
+        print url
+        conn = urllib2.urlopen(url)
```
Comment (nthomas-redhat) [inline tendrl/TimeSeriesDBManager.py]: We need to have single httpclient initailiazed for the application

File: tendrl/TimeSeriesDBManager.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    
+    @abstractmethod
+    def get_metrics(self, entity_name):
+        raise NotImplementedError()
+
+
+class GraphitePlugin(TimeSeriesDBPlugin):
+    def intialize(self):
+        configs = TendrlConfig()
+        self.host = configs.get('general', 'time_series_db_server')
```
Comment (nthomas-redhat) [inline tendrl/TimeSeriesDBManager.py]: Error conditions to be caught and act accordingly

File: tendrl/TimeSeriesDBManager.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        data = conn.read()
+        conn.close()
+        return data
+
+
+    def get_metrics(self, entity_name):
+        url = 'http://' + self.host + ':' + str(self.port) + '/metrics/index.json'
+        conn = urllib2.urlopen(url)
+        data = conn.read()
+        metrics = ast.literal_eval(data)
```
Comment (nthomas-redhat) [inline tendrl/TimeSeriesDBManager.py]: Handle the exceptions

File: tendrl/TimeSeriesDBManager.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        print "Need to implement metric fetcher for influxdb"
+
+    def get_metrics(self, entity_name):
+        print "Need implementation"
+
+
+@singleton
+class TimeSeriesDBManager():
+    def __init__(self):
+        self.time_series_db = TendrlConfig().get('general', 'time_series_db')
```
Comment (nthomas-redhat) [inline tendrl/TimeSeriesDBManager.py]: handle the error conditions

File: tendrl/utils.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
+import logging
+import os
+import ConfigParser
+
+
+class ConfigNotFound(Exception):
+    pass
+
+
+DEFAULT_CONFIG_PATH = "/etc/tendrl/tendrl.conf"
```
Comment (nthomas-redhat) [inline tendrl/utils.py]: Take this as input to the process rather than hardcoding it here

File: tendrl/utils.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+CONFIG_PATH_VAR = "TENDRL_CONFIG"
+
+
+def singleton(cls):
```
Comment (nthomas-redhat) [inline tendrl/utils.py]: Move this impl to a different file say singleton.py

File: tendrl/utils.py (hunk 12 above)
Comment (nthomas-redhat) [inline tendrl/utils.py]: +1

File: tendrl/Manager.py [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class TendrlPerformanceManager():
+    def __init__(self):
+        self.apiManager = api_manager.APIManager()
+        self.defaultsManager = defaults_manager.DefaultsManager()
+        self.defaultsManager.start()
```
Comment (anmolbabu) [inline tendrl/Manager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/APIManager.py (hunk 13 above)
Comment (anmolbabu) [inline tendrl/APIManager.py]: This had to be multiprocessing itslef as again the serveforecer used below blocks others

File: tendrl/APIManager.py (hunk 14 above)
Comment (anmolbabu) [inline tendrl/APIManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/TimeSeriesDBManager.py (hunk 17 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/TimeSeriesDBManager.py (hunk 19 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/utils.py (hunk 20 above)
Comment (anmolbabu) [inline tendrl/utils.py]: Removed now reusing the one from tendrl.common

File: tendrl/TimeSeriesDBManager.py (hunk 16 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/common/pull/64

File: tendrl/utils.py (hunk 21 above)
Comment (anmolbabu) [inline tendrl/utils.py]: Addressed in https://github.com/Tendrl/common/pull/64

**Comment by anmolbabu** [comment]
Accidentally closed by force rebase. The new PR is @ https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/utils.py (hunk 12 above)
Comment (anmolbabu) [inline tendrl/utils.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/1

File: tendrl/TimeSeriesDBManager.py (hunk 10 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/TimeSeriesDBManager.py (hunk 9 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/TimeSeriesDBManager.py (hunk 8 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/TimeSeriesDBManager.py (hunk 6 above)
Comment (anmolbabu) [inline tendrl/TimeSeriesDBManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/Manager.py (hunk 5 above)
Comment (anmolbabu) [inline tendrl/Manager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/APIManager.py (hunk 3 above)
Comment (anmolbabu) [inline tendrl/APIManager.py]: Its the server used internally by flask to serve apis. So stopping flask requires stopping this

File: tendrl/APIManager.py (hunk 2 above)
Comment (anmolbabu) [inline tendrl/APIManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: tendrl/APIManager.py (hunk 4 above)
Comment (anmolbabu) [inline tendrl/APIManager.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

File: setup.py (hunk 1 above)
Comment (anmolbabu) [inline setup.py]: Addressed in https://github.com/Tendrl/performance_monitoring/pull/2

**Pull request closed without merging**